File: src/test/java/dev/jlo/gamemodes/domain/common/MatchEventTest.java
package dev.jlo.gamemodes.domain.common;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class MatchEventTest {
    @Test
    void eventsOrderObjectiveThenScoreThenVictoryThenDeadline() {
        var events = List.of(
                new MatchEvent.DeadlineCheck(7),
                new MatchEvent.ScoreUpdate(7),
                new MatchEvent.Objective(7, 2),
                new MatchEvent.VictoryCheck(7),
                new MatchEvent.Objective(7, 1)
        );

        assertEquals(
                List.of(
                        new MatchEvent.Objective(7, 1),
                        new MatchEvent.Objective(7, 2),
                        new MatchEvent.ScoreUpdate(7),
                        new MatchEvent.VictoryCheck(7),
                        new MatchEvent.DeadlineCheck(7)
                ),
                events.stream().sorted(MatchEvent.ORDERING).toList()
        );
    }
}

File: src/test/kotlin/dev/jlo/gamemodes/domain/common/MatchEventTest.kt
package dev.jlo.gamemodes.domain.common

import kotlin.test.Test
import kotlin.test.assertEquals

class MatchEventTest {
    @Test
    fun `events order objective then score then victory then deadline`() {
        val events = listOf(
            MatchEvent.DeadlineCheck(tick = 7),
            MatchEvent.ScoreUpdate(tick = 7),
            MatchEvent.Objective(tick = 7, sequence = 2),
            MatchEvent.VictoryCheck(tick = 7),
            MatchEvent.Objective(tick = 7, sequence = 1)
        )

        assertEquals(
            listOf(
                MatchEvent.Objective(tick = 7, sequence = 1),
                MatchEvent.Objective(tick = 7, sequence = 2),
                MatchEvent.ScoreUpdate(tick = 7),
                MatchEvent.VictoryCheck(tick = 7),
                MatchEvent.DeadlineCheck(tick = 7)
            ),
            events.sortedWith(MatchEvent.ORDERING)
        )
    }
}
